csrc/scheduler/pointwise_utils.cpp

```diff
@@ -6,27 +6,109 @@
  */
 // clang-format on
 #include <ir_utils.h>
+#include <lower_utils.h>
 #include <scheduler/pointwise_utils.h>
+#include <utils.h>
 
 namespace nvfuser {
 namespace pointwise_utils {
 
-DomainMap::DomainMap(Fusion* fusion) : fusion_(fusion), ca_map_(fusion) {
-  tvs_with_rfactor_ = scheduler_utils::getTVsWithNonReductionRFactor(fusion);
-  for (auto select : ir_utils::getSelectOps(fusion)) {
-    select_ids_.emplace(select->getIndexedID());
-  }
-  for (auto select : ir_utils::getIndexSelectOps(fusion)) {
-    select_ids_.emplace(select->getIndexedID());
+namespace {
+
+// Grab all exact set pairs of producer and consumer domains of
+// indexed accesses, e.g., index_select
+std::vector<std::pair<
+    std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>,
+    std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>>
+getIndexedProducerConsumerPairs(Fusion* fusion, const ComputeAtMap& ca_map) {
+  std::vector<std::pair<
+      std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>,
+      std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>>
+      indexed_ids;
+
+  for (auto expr : fusion->exprs()) {
+    if (auto gather = dynamic_cast<TorchGatherOp*>(expr)) {
+      auto p_id = gather->getIndexedID();
+      auto c_id = gather->getConsumerOfIndexedID();
+      indexed_ids.emplace_back(
+          ca_map.disjointSetOf(p_id, IdMappingMode::EXACT),
+          ca_map.disjointSetOf(c_id, IdMappingMode::EXACT));
+    } else if (auto index_select = dynamic_cast<IndexSelectOp*>(expr)) {
+      auto p_id = index_select->getIndexedID();
+      auto c_id = index_select->getConsumerOfIndexedID();
+      indexed_ids.emplace_back(
+          ca_map.disjointSetOf(p_id, IdMappingMode::EXACT),
+          ca_map.disjointSetOf(c_id, IdMappingMode::EXACT));
+    } else {
+      // Note there's no consumer ID for select. This means we can't
+      // just propagate from consumers to indexed producers. It seems
+      // it's necessary to schedule producers and consumers separately
+      // in those cases.
+      continue;
+    }
   }
-  for (auto select : ir_utils::getTorchGatherOps(fusion)) {
-    TORCH_INTERNAL_ASSERT(select->input(0)->isA<TensorView>());
-    auto inp = select->input(0)->as<TensorView>();
-    auto inp_domain = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
-    for (auto id : inp_domain) {
-      select_ids_.emplace(id);
+
+  return indexed_ids;
+}
+
+// Check if a root ID of a fusion input tensor that is indirectly
+// accessed by ops such as torch_gather needs to be mapped with
+// a reference tensor. Select has a similar effect as squeeze as the
+// indexed domain is removed, so the domain does not need to be mapped
+// as long as the tensor is a fusion input. Similarly, in index_select
+// and torch_gather, if the output domain is a broadcast, it does not
+// need to be mapped if not resolved.
```

Comment:
So if I have
```c++
T2 = take_along_axis(lookup_tv=T0, index_tv=T1, axis=1)
T3 = squeeze(T2, {false, true})
```
where `T0` is `[2048, 100000000000000]`, and `T1` and `T2` are `[2048, 1]`, then can this be fused? I don't think so because we need to compute the entire `T0`, even if we just need one element of it. But the reference tensor does not have information about the second dimension of `T0`, so we will be unable to parallelize the computation.

Reply:
Note that this is what's done at line 43 of the TOT version of this file.

Reply:
Oh, if it is fusion input, then it's fine.

Reply:
Added an assertion